sycl/include/sycl/ext/oneapi/reduction.hpp

```diff
@@ -312,45 +270,460 @@ class reducer<T, BinaryOperation,
+              enable_if_t<Dims == 0 && View == true>>
+    : public combiner<reducer<T, BinaryOperation, Dims, Extent, Algorithm, View,
+                              enable_if_t<Dims == 0 && View == true>>> {
+public:
+  reducer(T &Ref, BinaryOperation BOp) : MElement(Ref), MBinaryOp(BOp) {}
+
+  void combine(const T &Partial) { MElement = MBinaryOp(MElement, Partial); }
+
+private:
+  T &MElement;
+  BinaryOperation MBinaryOp;
+};
+
+/// Specialization of 'reducer' class for array reductions exposing the
+/// subscript operator.
+template <typename T, class BinaryOperation, int Dims, size_t Extent,
+          class Algorithm, bool View>
+class reducer<T, BinaryOperation, Dims, Extent, Algorithm, View,
+              enable_if_t<Dims == 1 && View == false &&
+                          !IsKnownIdentityOp<T, BinaryOperation>::value>>
+    : public combiner<
+          reducer<T, BinaryOperation, Dims, Extent, Algorithm, View,
+                  enable_if_t<Dims == 1 && View == false &&
+                              !IsKnownIdentityOp<T, BinaryOperation>::value>>> {
+public:
+  reducer(const T &Identity, BinaryOperation BOp)
+      : MValue(Identity), MIdentity(Identity), MBinaryOp(BOp) {}
+
+  // SYCL 2020 revision 4 says this should be const, but this is a bug
+  // see https://github.com/KhronosGroup/SYCL-Docs/pull/252
+  reducer<T, BinaryOperation, Dims - 1, Extent, Algorithm, true>
+  operator[](size_t Index) {
+    return {MValue[Index], MBinaryOp};
+  }
+
+  T getIdentity() const { return MIdentity; }
+  T &getElement(size_t E) { return MValue[E]; }
+  const T &getElement(size_t E) const { return MValue[E]; }
+
+private:
+  marray<T, Extent> MValue;
+  const T MIdentity;
+  BinaryOperation MBinaryOp;
+};
+
+/// Specialization of 'reducer' class for array reductions accepting a span
+/// in cases where the identity value is known.
+template <typename T, class BinaryOperation, int Dims, size_t Extent,
+          class Algorithm, bool View>
+class reducer<T, BinaryOperation, Dims, Extent, Algorithm, View,
+              enable_if_t<Dims == 1 && View == false &&
+                          IsKnownIdentityOp<T, BinaryOperation>::value>>
+    : public combiner<
+          reducer<T, BinaryOperation, Dims, Extent, Algorithm, View,
+                  enable_if_t<Dims == 1 && View == false &&
+                              IsKnownIdentityOp<T, BinaryOperation>::value>>> {
+public:
+  reducer() : MValue(getIdentity()) {}
+  reducer(const T & /* Identity */, BinaryOperation) : MValue(getIdentity()) {}
+
+  // SYCL 2020 revision 4 says this should be const, but this is a bug
+  // see https://github.com/KhronosGroup/SYCL-Docs/pull/252
+  reducer<T, BinaryOperation, Dims - 1, Extent, Algorithm, true>
+  operator[](size_t Index) {
+    return {MValue[Index], BinaryOperation()};
+  }
+
+  template <typename _T = T, class _BinaryOperation = BinaryOperation>
+  static enable_if_t<has_known_identity_impl<_BinaryOperation, _T>::value, _T>
+  getIdentity() {
```

Comment:
I don't remember why the older code has that condition. It seems redundant. Can this be replaced with such code?
```suggestion
  T getIdentity() {
```